### PR TITLE
BUG: enable working around pickle compatibility issues on Py3 in npy files

### PR DESCRIPTION
**Pull request by pv**
Add pickle compatibility flags to numpy.save and numpy.load.  Allow only
combinations that cannot corrupt binary data in Numpy arrays.  Use the
same default values as Python pickle. These are necessary for enabling
Python 3 to load npy files produced on Python 2, when they contain object arrays.

In particular on the default values for `encoding=`:
- ASCII: can fail if the array contains binary data, or the serialized items contain non-ascii characters in strings
- latin1: can result to latin1 unicode representations of e.g. utf-8 strings (mojibake)
- bytes: results to bytes objects instead of strings in unpickled python objects
- all others: result to ndarray binary data corruption (eg object array containing other ndarrays)

All of this sucks, but I don't see ways out of this. `ASCII` is the default for pickle, and has the advantage of failing noisily rather than in obscure ways.

### CODE REVIEW

**Comment by pv** [comment]
It seems object arrays cannot be loaded at all from npz files on Py2.6, gh-5643, so the corresponding tests are skipped here.

File: numpy/lib/tests/test_format.py [hunk 1]
```diff
@@ -534,6 +534,69 @@ def test_python2_python3_interoperability():
...
+
+    expected = np.array([None, xrange, sixu('\u512a\u826f'),
+                         asbytes('\xe4\xb8\x8d\xe8\x89\xaf')],
+                        dtype=object)
+
+    for fname in ['py2-objarr.npy', 'py2-objarr.npz',
+                  'py3-objarr.npy', 'py3-objarr.npz']:
+        path = os.path.join(data_dir, fname)
+
+        if fname.endswith('.npz') and sys.version_info[0] == 2 and sys.version_info[1] < 7:
```
Comment (charris) [inline numpy/lib/tests/test_format.py]: Line is too long, also just below this.

**Comment by charris** [comment]
LGTM modulo long lines. Could you add something in the 1.10 release notes? Probably under improvements.

**Comment by pv** [comment]
Fixed

File: numpy/lib/format.py [hunk 2]
```diff
@@ -612,7 +621,16 @@ def read_array(fp):
...
-        array = pickle.load(fp)
+        if pickle_kwargs is None:
+            pickle_kwargs = {}
+        try:
+            array = pickle.load(fp, **pickle_kwargs)
+        except UnicodeError as err:
+            if sys.version_info[0] >= 3:
+                # Friendlier error message
+                raise UnicodeError("Unpickling a python object failed: %r\n"
+                                   "You may need to pass the encoding= option to numpy.load" % (err,))
```
Comment (charris) [inline numpy/lib/format.py]: PEP8 on this line also.

File: numpy/lib/npyio.py [hunk 3]
```diff
@@ -256,7 +261,7 @@ def __getitem__(self, key):
             bytes.close()
             if magic == format.MAGIC_PREFIX:
                 bytes = self.zip.open(key)
-                return format.read_array(bytes)
+                return format.read_array(bytes, pickle_kwargs=self.pickle_kwargs)
```
Comment (charris) [inline numpy/lib/npyio.py]: And this line also.

File: numpy/lib/npyio.py [hunk 4]
```diff
@@ -422,6 +459,11 @@ def save(file, arr):
         then the filename is unchanged.  If file is a string, a ``.npy``
         extension will be appended to the file name if it does not already
         have one.
+    fix_imports : bool, optional
+        Only useful in forcing objects in object arrays on Python 3 to be pickled
```
Comment (charris) [inline numpy/lib/npyio.py]: And these lines ;)

File: numpy/lib/npyio.py [hunk 5]
```diff
@@ -606,7 +654,7 @@ def _savez(file, args, kwds, compress):
             fname = key + '.npy'
             fid = open(tmpfile, 'wb')
             try:
-                format.write_array(fid, np.asanyarray(val))
+                format.write_array(fid, np.asanyarray(val), pickle_kwargs=pickle_kwargs)
```
Comment (charris) [inline numpy/lib/npyio.py]: And this line.

**Comment by charris** [comment]
I went all pep8 nitpicky on you ;)

**Comment by pv** [comment]
ok, folded

**Comment by charris** [comment]
Great. Thanks Pauli.

**Pull request merged**